server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -171,8 +159,7 @@ protected TransportReplicationAction(
         Writeable.Reader<Request> requestReader,
         Writeable.Reader<ReplicaRequest> replicaRequestReader,
         Executor executor,
-        boolean syncGlobalCheckpointAfterOperation,
-        boolean forceExecutionOnPrimary
+        ActionFlags... flags
```

Comment:
Instead of continuing to add booleans to this constructor, I thought it might be better to take a set of flags. This should help readability.

Reply:
I think it helps to show intention that there are some set of action flags, but it hides which flags exactly. Now there are 2 different places to look at. Also missing flag interprets as a false, but it might be a bug.

Personally, I find bare booleans in constructors/methods hard to read, need intellij support with inline hints (maybe github webui will them soon and it's no longer my problem :). I would prefer simple enums, also helps with static analysis. It does not take much space for enum, cant see any drawbacks.

```java
enum BypassCircuitBreakerOnReplica { Yes, No }
```

I dont even think all flags needs to be grouped by single enum. Other than they all booleans that change behaviour nothing in common. May be a single class that holds enums to reduce files clutter.

Reply:
My thinking was that the flags are deviations from the default (that's usually how they end up being added, and certainly the case for the one I added) so the absence of the flag just means "do the sensible default" or "do what it's always done".
Some advantages of this over having enums to replace the booleans
- don't have to specify them if you're happy with the default (I think that's a good thing)
- easy to add additional ones without going back to edit all the call sites
- don't need to define a bunch of boolean-proxy enums
- you can have defaults (we could do this by overloading the constructor but it's not as flexible and more verbose)

I think your suggestion makes more sense if there were no sensible default, and we wanted the caller to make that decision each time. But things like "bypass some safety check" (e.g. `ForceExecutionOnPrimary` and `BypassCircuitBreakerOnReplica`) don't fall into that category IMO.

But happy to go with consensus here. Ideally just want to avoid `true, false, true...`

Reply:
Are these flags indicate non-default behaviour? For example `TransportReplicationAction` and `GlobalCheckpointSyncAction` set `syncGlobalCheckpointAfterOperation` to `false`; and `TransportWriteAction` to `true`. Hard to tell what is default value, these are 2 different use cases. If these are truly 2 different use-cases and there is no default value I would prefer statically correct code, for example factory method that always set this flag to true or false. If it compiles it is correct. Or enums with assertions.

Reply:
I took the values used in the constructor that omits those flags to be the "default":

https://github.com/elastic/elasticsearch/blob/d8ff99748b5f2e7e27cc99a71578ffdca92bebb3/server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java#L131-L159

That constructor is used by the majority of direct descendants and it sets `syncGlobalCheckpointAfterOperation=false` and `forceExecutionOnPrimary=false`.

If you look back in history you can see these flags being added to the second constructor with the defaults/status quo behaviour added to the first constructor

`syncGlobalCheckpointAfterOperation`: f35d1de502a5c90384b0af5b2e9bb41d23041063
`forceExecutionOnPrimary`: c29cebb04deed28f2fd9f92e449450bd55b6369d

I think a builder is a good approach when there's a proliferation of constructor args, but because this is an abstract parent class, the only way to apply that pattern would be as a fluent-style parameters DTO passed to the parent class. Which seems like a significant refactor and one I'd probably not reach for just yet in this case.

Reply:
I think my preference here would be to change these boolean parameters into enums so that we can give the two options more specific names than just `true` and `false`. It is hard to read them at their call sites as things are today. I don't see a massive need for anything more complex than that. I would also be supportive of getting rid of the constructors which supply defaults for these parameters in favour of having each caller be explicit about the behaviour it wants. I don't see any huge advantage to hiding a couple of the 10+ parameters here, and experience shows it to be somewhat trappy.